### PR TITLE
Chore/metadata fields(need feedback)

### PR DESCRIPTION
**Pull request by eshadatta**
Hi @NYULibraries/hydra , @jgpawletko ,

Referencing PR #215 .  For context: This work is basically the implementation of the hackathon for models and vocabulary.rb(EDITED to clarify what I have implemented. The controller is being worked on by @da70 and @chrpr). I haven't changed the hardcoded fields in the rspecs(in fact I added a couple of fields in there to account for the slight addition of fields in the yml file) to demonstrate that all the tests are passing. I changed the yml file to reflect the structure @chrpr had put in which show the origin of the fields. I also added another hash to account for the namespace and the uri. I had to change the [controller](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/app/controllers/nyucores_controller.rb#L70) whitelisted params to all so that the tests would pass. We should probably have an argument that allows us to programmatically add whitelisted params rather than have it the original way, which was hardcoded. If you could also please review the additional [code](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/app/models/concerns/metadata_fields.rb#L30-L58) I had added that would be great. Also, if you could review the changed files in general, I'd really appreciate it. Thanks for your time.

I have rebased this against development, so it should be the most updated branch as of now. Thanks.

### CODE REVIEW

**Comment by eshadatta** [comment]
@NYULibraries/hydra, 
Per the code review feedback, I have implemented some of the suggestions but had questions about others. If you'd like to read the questions, please scroll down all the way. Otherwise, TL;DR..
 
I'm requesting feedback for the following files. If everyone could review them, I would really appreciate it.
1. [the metadata fields spec file](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/spec/models/concerns/metadata_fields_spec.rb)
2. [custom matcher](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/spec/support/custom_matcher.rb)
3. [module](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/app/models/concerns/metadata_fields.rb)

If you want to continue reading...my question is at the bottom.
The feedback was the following:
1. use the *_ALL variables when checking default value "all" in the metadata fields module
- DONE
2. split up allowed values function so that namespace and multiples can be tested separately...there are functions for that...and I had already started using those separately, so I'll go ahead and do that.
- DONE
3. the spec module should test all fields...broken down in the following:
- Test metadata module against a stub yaml file to check that it returns expected fields
- Test that it's returning the fields expected in the metadata fields yml file.

For # 3, I created a fixture of the metadata yaml file, basically just a copy of the yaml file and I'm testing that against the module. This is because the metadata fields module is using the rails [initializer](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/config/initializers/get_metadata_fields.rb#L1) to create a hash constant from the yaml file which is used everywhere in the code. So, I can't really use a stub yaml file to determine if it's reading a yaml file with the same structure and returning values. Instead, I've created a fixture of the metadata yml file and have basically used the same code in the rspec file and am testing that the expected fields are there. I'm a little confused as to how else I would test this. I also wrote a little custom matcher because I couldn't find a hash matcher for rspec2, there is one for rspec3.

**Comment by da70** [comment]
## metadata_fields_spec.rb

### General comments

* I think the testing method you've chosen makes sense.  Maybe we could use JSON format for the fixture as it would be easier to read and edit, and validation might a little more reliable because it's so much simpler than YAML.

* You mention: "the metadata fields module is using the rails initializer to create a hash constant from the yaml file which is used everywhere in the code".  At the moment it seems that only the module is using the `METADATA_FIELDS` hash from the initializer, no other code is.  But I just realized, a lot other code could later be using it, which means code could be written to bypass the module and access the data structure directly.  This breaks the encapsulation of the `MetadataFields` module which might not be desirable.

* Maybe we can more clearly indicate which values are coming from the spec methods, and which ones are coming from the code under test.  For example in the context starting at line 67, "check method returns based on various valid parameters", we could do these variable renames, which would make it less likely that a developer will switch them around by accident:
  * all -> all_fields_expected
  * all_fields -> all_fields_actual
  * fields -> all_fields_for_source_expected
  * fields_from_module -> all_fields_for_source_actual

* Good to see an example of how to make a custom matcher.  Is it worth thinking about upgrading to RSpec 3.0?

### process_metadata_fields

1) This test below will never indicate failure because if the test condition is false, the assertion will never run because the earlier `let(:all_fields)` will have error-ed out:

line 57:
```ruby
it { MetadataFields.should respond_to :process_metadata_fields }
```
...so we could probably just omit this test.

2) These two tests are checking something that was performed outside of them, namely the call to `MetadataFields.process_metadata_fields(...)` with no arguments in the top-level context `let(:all_fields)` assignments:

line 69:
```ruby
it 'returns an array if no arguments are sent' do
   expect(all_fields).to be_a_kind_of(Array)
end

it 'should equal all fields in the yml file if no arguments are sent' do
   all = get_fields_all_sources
   expect(all_fields).to match_array(all)
end
```

I think we should consider doing the actual `MetadataFields.process_metadata_fields` calls in the `it` statements themselves, as that eliminates the possibility that `all_fields` has been messed with between tests, or before these `it` statements are even reached.

Of course, we have been discussing a lot strategies for speeding up tests.  If we decide to bend the purist testing rules a bit and keep this var around, we might want to put it in the immediate `context` to keep it closer to the tests, and change the name from `:all_fields` to `:all_fields_returned_by_default_expected`.

3) Not sure if this is necessary:

line 101:
```ruby
context 'use of process_metadata_fields in an instance of a class' do
    subject { dummy_class.new }
        it { should respond_to(:process_metadata_fields) }
```

...because it it fails it would mean something core to Ruby was at fault: `Class.new(...)`, or `include`, or something like that.

### get_source_info

1) As with above, this test is run after a `let()` statement has already invoked this method:

```
    it { MetadataFields.should respond_to :get_source_info }
```

...so it can't ever indicate a failure, so we could omit this one, too.

**Comment by eshadatta** [comment]
@da70,

Thanks for the response. I'll respond to a few of the comments.

1. Good observation about the initializer. I can change the constant to a variable declared in the module rather than the initializer. But I don't want to pass the yaml file to the module, I rather it just be declared there and be read there. I'll skip the initializer part. This way it prevents the constant being used everywhere.
2. You seem to have a problem with YAML :-) but I don't want to change the fixture to a json representation because json can also be hard to read. The module and the test converts it to a json hash, so I don't see a problem with keeping it the way it is. Also, if the yaml is changed, the tests will fail, so again, I don't see a problem with keeping the yaml file. Also, the fixture is an exact copy of the yaml file being used, so it's a good way to test the module if there are additional fields or changes to the structure of the yaml file.
3. Your observation about being confused about where the data is coming from is totally right. Instead of renaming, I'm just going to call the module methods in the it statements.

I'll look through the rest of the rspec feedback soon.

Thanks.

**Comment by da70** [comment]
Yes I'm definitely YAML challenged! :)
The other reason for keeping it a separate format though is emphasize that those two files should be independently edited.  Sometimes there can be the temptation with fixture like this to just edit the real one, then copy it over to the test fixture to save time and effort, which defeats the purpose because it propagates errors into the test.
Also, if there is an error introduced into the real YAML file that is related to YAML syntax, e.g. incorrect indentation, there is a risk that the same error will be introduced into the fixture YAML file, even if it is independently edited.  It would not be possible to replicate a YAML syntax error in a JSON file.
It's like when checking sums of numbers on paper (before everyone started using calculators, spreadsheets, phones): what people sometimes did was the first time add the numbers going forwards, then to double-check, add the numbers going backwards.  The backup/verification is done differently than  real method.
If they are both YAML, though, I think it could work.  As long as we agree that we edit both files independently of each other.

**Comment by da70** [comment]
By the way, I wasn't saying that the initial copying of the real YAML file to fixture shouldn't have been done.  We are presuming our current YAML is correct, so the initial copy is absolutely appropriate.  I was referring to subsequent edits.  According to TDD if ever need to adjust the YAML file (for "type" and "format", for example), the procedure is to change the test first and make the edits to the fixture file, then let the test fail, and then edit the real YAML file and make the test pass.  The key would be to edit it, not simply copy the fixture over to the real or vice-versa.

**Comment by eshadatta** [comment]
@da70 ,

Yes, I think we should edit them independently once additions/edits are made to the "real" yaml file. I still think we should keep the fixture as a yaml since testing involves hewing to the real situation as closely as possible.

**Comment by da70** [comment]
Sure, keeping them both YAML seems fine.

**Comment by chrpr** [comment]
@eshadatta et al,

Just went through this. I like the approach, I like the use of a fixture and agree that it should remain a YAML file. I don't have much to add, especially on the rspec stuff -- reading through this code and @da70's comments has definitely been good for trying to learn rspec better, though!

I still would rather see the module live in /lib rather than /app/models/concerns. Is that something that I should change when I go through this? Also, thanks for working on the whitespace issues. But I think these should be two-space "soft-tabs" rather than "hard tabs", right? Or am I misreading the codestyle we use? https://github.com/bbatsov/ruby-style-guide#spaces-indentation If no one objects, I can do a whitespace cleanup (as a standalone commit) while I'm updating the PR I have out against this PR. 

That's all I have, other than to say thanks again to @eshadatta for her great work on this!

**Comment by da70** [comment]
Yes, thanks Esha, for doing all this heavy lifting, design work, and driving of important discussion!

**Comment by eshadatta** [comment]
@chrpr ,

Hi, yes, I agree the code should live in lib/, I just didn't change it because I figured you could move it when you worked on your branch. 

I seem to be really dumb with whitespace. I'll try and be cognizant of it. Thanks in advance for cleaning it up.

Also, thanks to you and @da70 for your work on this and  thanks everyone for your contributions. They're always very helpful.

**Comment by da70** [comment]
@eshadatta, FYI both Sublime or Atom should have an option for automatically converting a Tab keypress to spaces (2 or 4 as the case maybe).  That would probably be easiest, just set and forget.

**Comment by eshadatta** [comment]
@da70 ,

Yeah, I figured it out finally on Atom. Thanks.

**Comment by eshadatta** [comment]
@NYULibraries/hydra ,

I've incorporated some of the feedback. I've
1. Removed the initializer and moved the constant to the module
2. Changed some variable names

**Comment by jgpawletko** [comment]
Hi Esha,

Thank you for putting this code together.  Your code definitely
accomplishes the goal of gathering everything into one module. 
As we work toward a Gem, I think there are places we can refactor.
Whether we refactor now or later is up to you.

###### Some comments on the current code:
1. Ruby code indentation [should be 2 spaces](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)
2. could use chained assignment  
   `DEFAULT_NAMESPACE = NAMESPACE_ALL = 'all'`
3. should [freeze constants](http://blog.honeybadger.io/when-to-use-freeze-and-frozen-in-ruby/)  
   `DEFAULT_NAMESPACE = NAMESPACE_ALL = 'all'.freeze`
4. `MetadataFields.process_metadata_fields` feels like you're exposing implementation details in the method signature. Could you implement methods like the following?  
   `MetadataFields.single`  
   `MetadataFields.multiple`    
   `MetadataFields.dcterms`   
   ...   
  `MetadataFields.get_source_info` also seems like it's exposing implementation details...
5. `vocabulary.rb`  
   I noticed two `vocabulary.rb` files.  
   Do you need both?  
   `./lib/ichabod/vocabulary.rb`  
   `./lib/nyucore_metadata/vocabulary.rb`  

###### Some comments on a Gem (or refactoring the current code):
1. General code structure and complexity:  
   I'm a fan of [Sandi Metz](http://www.sandimetz.com/)  
   She put together some developer rules to manage code complexity.  
   [thoughtbot](https://thoughtbot.com/) put together an article on the rules [here](https://robots.thoughtbot.com/sandi-metz-rules-for-developers).  
   Bottom line, I think the `YAML` file hierarchy suggests some classes that might be useful in a Gem.  
2. Possible classes  
   The `YAML` file structure could guide the Gem classes.
   e.g.,   
    ```
   Terms      : has one or more Vocabulary objects
   Vocabulary : has an Info object, and multiple Field objects
   Info       : captures namespace and URI
                  (this class may not be needed.
                   could just add 'namespace' and 'uri' attributes to Vocabulary class)
   Field      : captures the name and attributes of the Field
    ```

3. Possible instance methods:  
    ```
    Terms#vocabularies  
                        # look at message patterns to see other methods needed.  
                        # e.g., should Vocabulary method results be cached and  
                        #       available here, e.g., Terms#single, Terms#multiple  
    ```
    ```
    Vocabulary#name      # could call this #label, or something else ...
    Vocabulary#info      # or could implement Vocabulary#namespace, Vocabulary#uri
    Vocabulary#fields
    Vocabulary#single
    Vocabulary#multiple
    ...                  # additional methods based on Attributes below
    ```

    ```
    Info#namespace
    Info#uri
    (this class might not be needed, should look at message interactions...)
    ```

    ```
    Field#name          # could call this #label, or something else ...
    Field#data_type   
    Field#facetable    
    Field#indexed      
    Field#mandatory    
    Field#multiple     
    Field#searchable   
    Field#stored  
    ```

4. Each `Vocabulary` could have it's own `YAML` file.
   During initialization, `Ichabod` could load all of the vocabulary files
   in a given directory into a set of Vocabulary objects that are assigned
   to a `Terms` object.

##### Going forward
We should diagram the message interactions between `NYUCore` (or `Vocabulary`?)
Gem objects and Ichabod to see if the classes should be restructured,
e.g., should `Terms` respond to `#single` and `#multiple`, should `Field` have an `#attributes` method, and that returns an array of `Attribute` objects. If so, what does the `Attribute` class look like...

I hope this feedback is useful...
best-
Joe

**Comment by da70** [comment]
FYI, regarding:

> 3. should [freeze constants](http://blog.honeybadger.io/when-to-use-freeze-and-frozen-in-ruby/)  
>    `DEFAULT_NAMESPACE = NAMESPACE_ALL = 'all'.freeze`

Note that this method creates an immutable string and assigns it to the constant(s), but this doesn't prevent the constants from being changed by changing them to refer to completely different strings via reassignment:

```Ruby
david@GURGI:~/> irb
2.1.3 :001 > FROZEN_CONSTANT = "This is a frozen string.  It cannot be altered!".freeze
 => "This is a frozen string.  It cannot be altered!" 
2.1.3 :002 > FROZEN_CONSTANT << '  This will not be added.'
RuntimeError: can't modify frozen String
	from (irb):2
	from /Users/david/.rvm/rubies/ruby-2.1.3/bin/irb:11:in `<main>'
2.1.3 :003 > FROZEN_CONSTANT = 'But the original frozen string can be completely replaced.'
(irb):3: warning: already initialized constant FROZEN_CONSTANT
(irb):1: warning: previous definition of FROZEN_CONSTANT was here
 => "But the original frozen string can be completely replaced." 
2.1.3 :004 > FROZEN_CONSTANT
 => "But the original frozen string can be completely replaced." 
2.1.3 :005 > 
``` 

In fact, a non-constant var exhibits the same behavior:

```Ruby
2.1.3 :001 > non_constant_var = "This is a frozen string.  It cannot be altered!".freeze
 => "This is a frozen string.  It cannot be altered!" 
2.1.3 :002 > non_constant_var <<  '  This will not be added.'
RuntimeError: can't modify frozen String
	from (irb):2
	from /Users/david/.rvm/rubies/ruby-2.1.3/bin/irb:11:in `<main>'
```

I think it's a good idea to freeze the strings for our constants, but we should just be aware that Ruby constants unfortunately are not truly constant, they seem basically to be a syntactic constructs that trigger warnings.

Have only skimmed Joe's feedback.  Lots of food for thought here, thanks Joe.  I'm going to check out the Sandi Metz article.

**Comment by chrpr** [comment]
Hi David, Joe, et al.,

Hey it's the source of my comment about "methods should only be 5 lines"! :)

I knew that principle had come from somewhere authoritative.

Thanks for sharing this, Joe! Super helpful to review some of these
principles, and also to know when it's wise to break them.

-Corey

On Mon, Apr 11, 2016 at 6:21 PM, da70 <notifications@github.com> wrote:

> FYI, regarding:
>
>
>    1. should freeze constants
>    <http://blog.honeybadger.io/when-to-use-freeze-and-frozen-in-ruby/>
>    DEFAULT_NAMESPACE = NAMESPACE_ALL = 'all'.freeze
>
> Note that this method creates an immutable string and assigns it to the
> constant(s), but this doesn't prevent the constants from being changed by
> changing them to refer to completely different strings via reassignment:
>
> david@GURGI:~/> irb2.1.3 :001 > FROZEN_CONSTANT = "This is a frozen string.  It cannot be altered!".freeze => "This is a frozen string.  It cannot be altered!" 2.1.3 :002 > FROZEN_CONSTANT << '  This will not be added.'RuntimeError: can't modify frozen String    from (irb):2    from /Users/david/.rvm/rubies/ruby-2.1.3/bin/irb:11:in `<main>'2.1.3 :003 > FROZEN_CONSTANT = 'But the original frozen string can be completely replaced.'(irb):3: warning: already initialized constant FROZEN_CONSTANT(irb):1: warning: previous definition of FROZEN_CONSTANT was here => "But the original frozen string can be completely replaced." 2.1.3 :004 > FROZEN_CONSTANT => "But the original frozen string can be completely replaced." 2.1.3 :005 >
>
> In fact, a non-constant var exhibits the same behavior:
>
> 2.1.3 :001 > non_constant_var = "This is a frozen string.  It cannot be altered!".freeze
>  => "This is a frozen string.  It cannot be altered!" 2.1.3 :002 > non_constant_var <<  '  This will not be added.'RuntimeError: can't modify frozen String    from (irb):2    from /Users/david/.rvm/rubies/ruby-2.1.3/bin/irb:11:in `<main>'
>
> I think it's a good idea to freeze the strings for our constants, but we
> should just be aware that Ruby constants unfortunately are not truly
> constant, they seem basically to be a syntactic constructs that trigger
> warnings.
>
> Have only skimmed Joe's feedback. Lots of food for thought here, thanks
> Joe. I'm going to check out the Sandi Metz article.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/NYULibraries/ichabod/pull/221#issuecomment-208591164>
>



-- 
Corey A Harper
Metadata Services Librarian
New York University Libraries
20 Cooper Square, 3rd Floor
New York, NY 10003-7112
212.998.2479
corey.harper@nyu.edu

**Comment by eshadatta** [comment]
@jgpawletko ,

Thanks for the very helpful feedback. Yes, I definitely would like to convert the YAML into classes and instance methods when we move fortward to a gem. In fact, eventually, instead of a YAML file, we would benefit in having a front end so that we wouldn't depend on a yaml file for structure and metadata experts can enter metadata fields details as they see fit. I figured I wouldn't here since it was a stop gap before a really well-architected solution was done in the gem. I'll go ahead and change the module methods to remove implementation details in the signature and also freeze the constants. Thanks again!

**Comment by da70** [comment]
@chrpr, yes I thought of you when I read that! :)  I agree with the rule in principle.  For me I tend to keep in mind "keep methods short, single-purposed, easy to understand, when appropriate extract code into separate methods for testability, and for abstracting out details that don't need to be kept in mind while trying to understand what the current method is doing".

A rule like "5 lines" I think might possibly be a good guideline on occasion.  Particularly if there are static analysis tools like [PMD](https://pmd.github.io/pmd-5.4.1/index.html) or [PHPMD](https://phpmd.org/rules/index.html) being used to control complexity on extremely large codebases that no one has the time to analyze in detail by eye, it can be good to quantify rules so that machines can give hints as to which areas of the code should be looked at.  I've used PHPMD before, more for fun than anything else, and I did like looking at the data it provided.  Incidentally, I notice PHPMD doesn't report a problem unless [a method exceeds 100 lines](https://phpmd.org/rules/codesize.html#excessivemethodlength).

The key for me is "guideline".  Sometimes when these rules are expressed with a hard number people can become fixated on that and assume that the rule is more absolute than it is, and it becomes overly mechanical.  Often it is appropriate to bend or break the rules -- the thoughtbot article Joe linked to does say this.  
Another useful authoritative source: ["guideline or rule"](https://www.youtube.com/watch?v=ux8nTqyB6Bk).

**Comment by da70** [comment]
@eshadatta, @jgpawletko 
How is the current `MetadataFields.process_metadata_fields` param list exposing implementation details, exactly?  Is it because the first param is called "namespace" instead of "schema"?  It doesn't seem to actually tie us to YAML implementation to have namespace and single/multiple params.

Also, as I understand it, to break out this method into many with params removed would require all permutations:

`MetadataFields.single_dcterms`
`MetadataFields.multiple_dcterms`
`MetadataFields.single_ichabod`
`MetadataFields.multiple_ichabod`
`MetadataFields.single_rdf_dc`
`MetadataFields.multiple_rdc_dc`
....etc.

It seems to be a lot of wrappers, and every time the user of the gem wants to add new schema/namespace more methods would have to be written.  We are also locking them into our particular set of namespaces by having method names that have the namespaces hardcoded in them.  Is it our intention to require `dcterms` and other namespaces?
Maybe this is addressed by what Joe wrote later about each Vocabulary having its own YAML file?  (I don't actually know what Vocabulary refers to in this context.  Have not yet had time to delve into this)

**Comment by jgpawletko** [comment]
Hi @da70 ,

Thank you for your comments.

I might be overthinking the method signatures, and others should weigh in, especially when we start work on a Gem. 

Re: having a bunch of method permutations: dynamic programming can help here, (not super-sophisticated, inscrutable dynamic programming, just simple [`#define_method`](http://ruby-doc.org/core-2.2.0/Module.html#method-i-define_method) stuff).  The methods could be defined based on the contents of the `YAML` files.  

If we were to go the dynamic programming route, you could have the following:
`MetadataFields.dcterms.single` which returns all single-valued `Fields` in the `dcterms` `Vocabulary`
`MetadataFields.ichabod.multiple` which returns all multiple-valued `Fields` in the `ichabod` `Vocabulary`
etc.

This is all speculation, however.

I think we should write up interaction diagrams first to see what messages actually need to be sent between objects, then worry about the implementation strategy.   My comments/proposals re: possible classes and methods were meant to stimulate discussion around refactoring.

It may be that the method signatures that Esha developed are the best approach, which would be great, as they're already implemented. :smile: 

cheers...

**Comment by eshadatta** [comment]
@jgpawletko, @da70 ,

Yeah, as I was looking into this, I had similar questions that David had. I figured metaprogramming would be the best bet since implementing methods per source and occurrence will be brittle.
TL;DR: I'll devote a day or two to develop a proof of concept with Struct and see if it's feasible. Regardless, it will result in a fair amount of re-writing of code and tests. I'll also ask @NYULibraries/hydra  at scrum to see if an extensive re-write of code is merited right now. Basically, it involves iterating through hashes of various permutations of the yml data to create the methods needed.

I tried a few approaches this morning for the past 2 hours, so it's not super informed:
1. Use **define_method**: this fails because the stack level is too deep if I just pass the hash of the yaml file as is. Also, need to investigate further exactly how to use it
2. Use **Open Struct**: this creates instance methods but also creates write methods which is something we absolutely don't want
3. Use **Struct** - need to investigate. Seems promising. It's pretty performant and strict and we can define methods as we see fit
4. Use **method_missing**: I feel uncomfortable using this. I don't really like to overload this to define methods even though it's apparently common practice.

**Comment by eshadatta** [comment]
@jgpawletko , @da70 ,

I do like these discussions, so thanks to you both and everyone else for the feedback. I really appreciate the time everyone has taken for this!

**Comment by da70** [comment]
@jgpawletko, @eshadatta 

I think `MetadataFields.dcterms.single` and `MetadataFields.ichabod.multiple` type methods would make for a very clean and intuitive module interface.
Esha seems to have gotten pretty far exploring dynamic programming options.  I agree we probably should avoid using `method_missing` for this.

As for how much time we spend, we could do it in stages: first write out the wrapper methods with hardcoded names to get the cleaner module interface, then later change to dynamic generation if we can prove that it eventually would be feasible.

**Comment by eshadatta** [comment]
@da70 ,

I'm reluctant to go down the path of hardcoded wrapper names. I'd rather it be left like this and then be re-architected, since that's going to happen anyways. I'll look more into metaprogramming because that'll be helpful for the gem.

**Comment by da70** [comment]
@eshadatta, sounds good.  We learn a lot that way, too.

**Comment by jgpawletko** [comment]
@eshadatta ,
If we decide to implement the classes I mentioned, then I think the only bit of dynamic programming required is to add method to `MetadataFields` module for each `Vocabulary#name`.
Each `Vocabulary` object would support the `#multiple`, `#single`, etc. methods.

So, the dynamically created method would return the vocabulary object that matched that name.
You could use a `Hash` in the `Terms` class to store the `Vocabulary` objects, and then the dynamically programmed accessor method e.g., `MetadataFields.nyucore` would just use the vocabulary name as the key into the `Hash`.

We can meet about this, or I can mock up the classes.
I still think we should understand what we need these objects to do via an interaction diagram.

Best-
Joe

**Comment by eshadatta** [comment]
@jgpawletko ,

Hi Joe,

Yup, it would be great to meet. Just sent you an email. I have some working code but I'd like to talk through it before continuing on just so that we're on the same page?

Thanks!

**Comment by jgpawletko** [comment]
Hi @eshadatta ,

I was looking a little deeper at the code to understand the object interactions.
One question: is your code using the `facetable` information from the YAML file?
I ask because [this](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/lib/ichabod/nyucore_datastream.rb#L11) doesn't access the `MetadataFields` module.

Thanks!

**Comment by eshadatta** [comment]
Hi @jgpawletko ,

My code right now isn't doing anything with the facetable attribute in the yml file. I didn't change that [line](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/lib/ichabod/nyucore_datastream.rb#L11) of code because I didn't understand what exactly [this line](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/lib/ichabod/nyucore_datastream.rb#L21) is doing. I had brought it up at the code review but didn't mention it in the summary email I had sent. We had decided to investigate this code [segment](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/lib/ichabod/nyucore_datastream.rb#L18-L26) deeper at some tech meeting. I figured I'd cycle back to it once it became clearer. Thanks for checking. I'll come up with an interaction diagram soon.

**Comment by jgpawletko** [comment]
Thanks, @eshadatta .

My understanding is that this code
```
index_args << :facetable if FACETABLE_TERMS.include? term
```
controls how a `term` is indexed, i.e., if the current `term` in the iteration 
is in the `FACETABLE_TERMS` array, then that information is passed 
to the indexer by including the `:facetable` symbol in the `index_args` array.

The facetable term, are:
```
FACETABLE_TERMS = [:creator, :type, :language, :subject]
```

:+1: for digging into the code block you highlighted.

The reason I'm asking about this is after adding my initial comments to this PR I wanted to
learn more about how to dynamically program methods based on keys in the YAML
file.

I started testing out approaches in [a separate gem](https://github.com/jgpawletko/tvf) because the Ichabod test suite takes so long to execute.

I learned that I could use `define_singleton_method` to dynamically add methods with the
names of keys in the YAML file, e.g., [this line](https://github.com/jgpawletko/tvf/blob/master/lib/tvf/vocabulary.rb#L31).

I have several classes working, but I didn't want to go further without the 
interaction diagram analysis.  For instance, the gem can support this pattern: `terms.dcterms.facetable`, and `terms.dcterms.multiple`, but there's no reason to add
 `#facetable` and `#multiple` if they aren't needed by Ichabod.

Thanks for your time!
Joe

**Comment by eshadatta** [comment]
hi @jgpawletko ,

RE: singleton methods, awesome! or shall I say 👍  :-)

Sorry, I wasn't clear about my confusion. 
I realized that `index_args << :facetable if FACETABLE_TERMS.include? term` included terms to be faceted on but I don't know what it actually does? The facets list in ichabod is also defined [here](https://github.com/NYULibraries/ichabod/blob/ab414a4a563e87f88c0c43e01b476c4d4533b87d/app/controllers/catalog_controller.rb#L46-L50) in the catalog controller. So...I'm not sure of that code is actually doing anything or if it's doing something at the Fedora level because the class inherits from ActiveFedora::NtriplesRDFDatastream? And what is it doing there? Which is why I was going to investigate further.

multiple is being used in ichabod. Right now, this [code](https://github.com/NYULibraries/ichabod/blob/chore/metadata_fields/app/models/nyucore.rb#L5-L6) in models/nyucore.rb needs all fields from all sources that occur only once and occur multiple times to populate to constants that are then added to the two metadata streams, source and native. So, the application will need methods that will get all fields that are multiple and all fields that are single occurrences, regardless of the source...like ichabod or dcterms or whatever.

**Comment by eshadatta** [comment]
@jgpawletko , @NYULibraries/hydra ,

I had a meeting with Joe and we decided that we would keep the approach in this PR as is and re-visit once the work on the gem is started. Should this be merged?

**Comment by jgpawletko** [comment]
@eshadatta , @NYULibraries/hydra 
You highlighted some key questions in your comments re: multiple definitions of facets.  I'm hesitant to merge in code when we don't fully understand the impact.  Could we discuss the code in the Ichabod Tech meeting scheduled for this Thursday?
Please let me know your thoughts.
Thanks!

**Comment by chrpr** [comment]
@jgpawletko @eshadatta @NYULibraries/hydra 

I'm not sure we need to hold off on merging this, given that the PR hasn't actually changed any of that facetable code. Those lines are the same as they were previously, and we can iterate on hooking that chunk of things into the YAML, right?

In any case, I think I can explain facetable at the ActiveTriples Datastream level vs. Facets in the catalog controller. The code we're looking at now is used by active_fedora's to_solr method. Remember all the dynamic field support that Hydra has baked into the solr schema. Declaring something facetable tells hydra that on to_solr, [solrizer ](https://github.com/projecthydra/solrizer/blob/d6b2e2d65041a0d204b4541556c72298b227fca4/README.md)should create a dynamic copy field typed foo_sim. This makes sure SOLR can facet on this field.

The controller code, on the other hand, is telling _blacklight_ what facets to actually render in the UI. There's not necessarily anything that says these have to match. We could make sure solr has a *_sim copy of a field, and not actually facet on it in the UI. I doubt we could facet in the UI if there weren't an appropriate solr copy field, though.

Let me know if you have any questions about that! And thanks again for your hard work on this. I'm still trying to wrap my head around the rest of this convo, but I thought I could at least help out here. :)

-Corey

**Comment by eshadatta** [comment]
@chrpr ,

Awesome, thanks for the great explanation! @jgpawletko , if you could please merge when you have a second, that would be great. @ekate has a PR open and will have to rebase this code to her branch.

Thanks.
